pkg/insight/insightstore/deploymentstore.go

```diff
@@ -17,26 +17,391 @@ package insightstore
+		return nil, nil, err
+	}
+	if size > maxChunkByteSize {
+		return nil, nil, errExceedMaxSize
+	}
+
+	// Create meta
+	newChunkMetaData := model.InsightDeploymentChunkMetadata_ChunkMeta{
+		From:  from,
+		To:    to,
+		Name:  key,
+		Size:  size,
+		Count: int64(len(deployments)),
+	}
+
+	meta := &model.InsightDeploymentChunkMetadata{
+		Chunks:    []*model.InsightDeploymentChunkMetadata_ChunkMeta{&newChunkMetaData},
+		CreatedAt: updatedAt.Unix(),
+		UpdatedAt: updatedAt.Unix(),
+	}
+
+	return meta, chunkData, nil
+
+}
+
+// deployments must not be empty
+func createNewChunkAndUpdateMeta(curMeta *model.InsightDeploymentChunkMetadata, deployments []*model.InsightDeployment, updatedAt time.Time) (*model.InsightDeploymentChunk, error) {
+	newChunk, size, err := createNewChunk(deployments)
+	if err != nil {
+		return nil, err
+	}
+	if size > maxChunkByteSize {
+		return nil, errExceedMaxSize
+	}
+
+	// Create new meta and chunk
+	newChunkKey := determineDeploymentChunkKey(len(curMeta.Chunks) + 1)
+
+	// Create meta
+	from, to := deployments[0].StartedAt, deployments[len(deployments)-1].StartedAt
+	newChunkMetaData := model.InsightDeploymentChunkMetadata_ChunkMeta{
+		From:  from,
+		To:    to,
+		Name:  newChunkKey,
+		Size:  size,
+		Count: int64(len(deployments)),
+	}
+
+	curMeta.Chunks = append(curMeta.Chunks, &newChunkMetaData)
+	curMeta.UpdatedAt = updatedAt.Unix()
+
+	return newChunk, nil
+}
+
+// deployments must not be empty
+func appendChunkAndUpdateMeta(meta *model.InsightDeploymentChunkMetadata, curChunk *model.InsightDeploymentChunk, deployments []*model.InsightDeployment, updatedAt time.Time) error {
+	firstDeployment := deployments[0]
+	lastDeployment := deployments[len(deployments)-1]
+	if firstDeployment.StartedAt < curChunk.To {
+		return errInvalidArg
+	}
+
+	curChunk.Deployments = append(curChunk.Deployments, deployments...)
+	curChunk.To = lastDeployment.StartedAt
+
+	size, err := messageSize(curChunk)
+	if err != nil {
+		return err
+	}
+
+	// we cannot know exact size on caller side.
+	// so maybe this condition become true.
+	// this is safe if we think maxChunkByteSize as soft limit
+	// if size > maxChunkByteSize {
+	// 	return errExceedMaxSize
+	// }
+
+	latestChunkMeta := meta.Chunks[len(meta.Chunks)-1]
+
+	// Update meta
+	latestChunkMeta.Size = size
+	latestChunkMeta.To = lastDeployment.StartedAt
+	latestChunkMeta.Count += int64(len(curChunk.Deployments))
+	meta.UpdatedAt = updatedAt.Unix()
+
+	return nil
+}
+
+// Load proto message stored in path. If path does not exists, dest does not modified.
+func (s *store) loadProtoMessage(ctx context.Context, path string, dest proto.Message) error {
+	raw, err := s.filestore.Get(ctx, path)
+	if err != nil {
+		return err
+	}
+
+	err = proto.Unmarshal(raw, dest)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *store) storeProtoMessage(ctx context.Context, path string, data proto.Message) (dataSize int64, err error) {
+	raw, err := proto.Marshal(data)
```

Comment:
Proto has a smaller size but  JSON has better readability and portability.
So I prefer JSON to Proto.

Reply:
Want to hear thoughts from other members.

Reply:
I agree🙏

Reply:
🙆‍♀️